game-app/game-core/src/main/java/games/strategy/engine/random/PlainRandomSource.java

```diff
@@ -20,9 +23,37 @@
     checkArgument(max > 0, "max must be > 0 (%s)", annotation);
     checkArgument(count > 0, "count must be > 0 (%s)", annotation);
 
-    final int[] numbers = new int[count];
-    for (int i = 0; i < count; i++) {
-      numbers[i] = getRandom(max, annotation);
+    final int[] numbers = new int[count]; // initialize return array of drawn numbers
+    if (count > 2) {
+      // get base power results (compute if needed)
+      BasePowers basePowers = basePowersMap.get(max);
+      if (basePowers == null) {
+        basePowers =
+            new BasePowers((int) Math.abs(Math.log(Integer.MAX_VALUE) / Math.log(max)), max);
+        basePowersMap.put(max, basePowers);
+      }
+      final int basePowerMax = basePowers.getBasePowerMax();
+      int restCount = count;
+      int diceIdBase = 0;
+      while (restCount > 0) {
+        final int powerCur = Math.min(basePowerMax, restCount); // = random numbers to draw now
+        int randomInt = getRandom(basePowers.getPowerResult(powerCur), annotation);
+        // Get individual dice x by projecting from range 1 until base^x to 0 until (base-1) for
+        // each x via modular arithmetic from the largest power to the smallest
+        // Note: x is diceIdOffset
+        for (int diceIdOffset = powerCur - 1; diceIdOffset >= 0; --diceIdOffset) {
+          final int powerResult = basePowers.getPowerResult(diceIdOffset);
+          final int powerFit = randomInt / powerResult;
+          numbers[diceIdBase + diceIdOffset] = powerFit;
+          randomInt -= powerFit * powerResult;
+        }
+        diceIdBase += powerCur;
+        restCount -= powerCur; // newly remaining random numbers to be drawn
+      }
```

Comment:
Is there a benchmark that shows that this is faster? Given there's a bunch of extra logic and complexity...

Or is the main benefit that getRandom() may use the network and we reduce the number of network calls? If so, that can be improved by introducing a getRandomNumbers(count) API that can return a bunch of random numbers...

Reply:
Check out the test that has been added. It is faster even without network benefits

Reply:
I tried it on my machine (M1 MacBook Air) and I'm not seeing conclusive results.

For example, I did two runs (one with the new code and another with the if changed to always be false) and modified the test to print endTime - startTime and got the following:
37960750
37266916

These seem close enough that it doesn't seem like the extra complexity might be worth it?

Reply:
I modified the test to run both versions (via a parameter on the function) and added an outer loop to do it 100 times (measuring still the inner loop time). This way, we can get JIT to stabilize. Here's what I'm seeing for the last few runs:

Normal 1966334
Batch 1999584
Normal 2012250
Batch 2044959
Normal 1965208
Batch 2128500
Normal 1959833
Batch 2030709
Normal 2017333
Batch 2052708
Normal 2049250
Batch 2082334
Normal 2134750
Batch 2064209
Normal 2134000

So it's a bit of a toss up.

Here's my changes:
https://pastebin.pl/view/raw/26feb56b

Are your results different? I'm curious if running on x86 shows something else.

Reply:
I also wonder if some of the inner loop logic can be modified to speed up your new code. For example, using bit operators and avoiding division and multiplication (e.g. just using bit masks / shifts).

Reply:
@asvitkine I did the test measurements outside the method (with parameter switch) from a test method multiple times and mostly got for 5-45% performance improvement, but a few time the old way was faster. My assumption was that the time measurement was affected by other programs running.
I agree with @DanVanAtta , that we should test carefully, but do not know how to do that "correctly". The new Java Profiler might be a good start, but we might have to find an agreement between each other what is a valid test first. It does not make sense if I do tests and everyone else has to rebuild and repeat them. Is there the option of test cases that are not regularly executed with the other test cases, but only for performance tests?

Reply:
There is a PerfTimer class that can be used to time methods or specific lines of code. That and/or good old fashioned stop watch (or wall clock). To check this, set up a large battle and then start checking for time differences.

I like the idea of this change, but we do need to be able to concretely the benefits for the "extra" complexity. For example, the performance test that was written, how does that compare with the old algorithm vs the new?

As far as memory usage, I suspect it is not significant. TripleA keeps a full res image of the entire map in memory just to shrink it down & render the minimap - which is in addition to the memory used to actually render the map. So all that is to say is that we can mostly focus on wall-clock time for performance measurements.

Reply:
>  how does that compare with the old algorithm vs the new?

Sorry, I realize @asvitkine did exactly this. I wonder where the difference is, can we run this test in some sort of setup that shows a strong difference?

Reply:
2 Result (New vs. Old): 35 : 15
3 Result (New vs. Old): 48 : 2
4 Result (New vs. Old): 46 : 4
5 Result (New vs. Old): 48 : 2
6 Result (New vs. Old): 49 : 1
7 Result (New vs. Old): 50 : 0
8 Result (New vs. Old): 50 : 0
9 Result (New vs. Old): 50 : 0
10 Result (New vs. Old): 50 : 0
11 Result (New vs. Old): 50 : 0
12 Result (New vs. Old): 50 : 0
13 Result (New vs. Old): 50 : 0
14 Result (New vs. Old): 50 : 0
15 Result (New vs. Old): 50 : 0
16 Result (New vs. Old): 50 : 0
17 Result (New vs. Old): 50 : 0
18 Result (New vs. Old): 50 : 0
19 Result (New vs. Old): 50 : 0

Each line is a test for x (first number) number of random numbers that was run 100 times (to get a better average time) and compared with 50 times between the new enhanced version and the old, being a simple call to the method `public int getRandom(final int max, final String annotation)`. The ratio between the two is shown at the end of each line.
I deduce from the output that I pasted above that the new enhanced version is better on average in all cases and gaining complete dominance over the old version when the number of random numbers is high (increasing dominance the more numbers are calculated).

The overall runtime of this test was 283ms (for ca. 20*100*2*50 = 400.000 random numbers in total) and therefore I doubt that drawing number is a general performance issue. The reason why I started this PR was a theory that the new version should be faster. This is true, but the effect seems to be small. 
I would therefore suggest to close this PR discussion as 

1. we have bigger fishes to fry (performance gain is small)
2. the results show a benefit and the increased complexity is acceptable from my point of view

@DanVanAtta, @asvitkine: What do you guys say?

Reply:
Can you share the code you used to test to produce the above output? I'm not exactly following - for example, I don't understand what a "ratio" of "50 : 0" means.

Reply:
Thanks! I tried your code and the results are similar for me.

I also did a sum of the two runtimes to see not just how often the new way is faster, but also what's the actual difference.
I'm seeing about a 60% improvement of the new algorithm. Of course, this is over the specific set of dice amounts being measured by the test - which is probably not representative of a real game.

However, the dice rolling is very fast either way. I had to bump the `testRuns` value to 1000 to get even to ~1/10 exec time for each of the algorithms. So with 1000 `testRuns`, we roll 18,900,000 dice total, half using the old algo and half using the new, and the total runtime for the old algo is 150ms vs. 96ms with the new.

So even with the old algo, it's able to roll 126 million dice / sec.

So from the above, my vote is to abandon this change, because dice rolling is already very fast and the extra complexity is not worth the improvement in speed in this area.  (Although it may be interesting to check how many dice are rolled in in practice for an AI-heavy game where there's lots of simulations being done - though that's also multithreaded.)

As an aside, if we did want to pursue this, I would want us to write some robust tests to make sure the new algorithm is producing fair results - per my other comment thread (I haven't replied with a suggestion on how since I wanted this to be resolved. As to how, I could imagine something like rolling dice and doing a tally of the number of times each face comes up and then doing a Chi-Square test to determine if they're fair.)

@DanVanAtta do you agree with my thoughts above?